### PR TITLE
Added encoder and decoder modules for RETRO model

### PR DESCRIPTION
**Pull request by yidong72**
# What does this PR do ?

Add Rotary positional embedding to parallel attention 
Add Chunked Cross attention module
Add RETRO encoder module
Add RETRO decoder module
Add unit-test for individual modules.

### CODE REVIEW

File: nemo/collections/nlp/modules/common/megatron/language_model.py
Comment (ericharper) [inline nemo/collections/nlp/modules/common/megatron/language_model.py:175]: Please update the docstring with this arg

**Review by MaximumEntropy — CHANGES_REQUESTED** [review]
Really nice PR :) I loved the unittests. A few comments for now, will do another pass in a bit.

File: nemo/collections/nlp/modules/common/megatron/transformer.py
Comment (ericharper) [inline nemo/collections/nlp/modules/common/megatron/transformer.py:392]: Can we call this `rotary_pos_emb` ?

File: nemo/collections/nlp/modules/common/megatron/transformer.py
Comment (ericharper) [inline nemo/collections/nlp/modules/common/megatron/transformer.py:1282]: Is it possible to do a stronger check here?

File: nemo/collections/nlp/modules/common/megatron/retrieval_transformer.py
Comment (yidong72) [reply nemo/collections/nlp/modules/common/megatron/retrieval_transformer.py:124]: good point. it should use `kv_channel`

File: nemo/collections/nlp/modules/common/megatron/retrieval_transformer.py
Comment (yidong72) [reply nemo/collections/nlp/modules/common/megatron/retrieval_transformer.py:162]: good idea. I change it not to pass in the device.

File: nemo/collections/nlp/modules/common/megatron/retrieval_transformer.py
Comment (yidong72) [reply nemo/collections/nlp/modules/common/megatron/retrieval_transformer.py:227]: yes. I changed the name.

File: nemo/collections/nlp/modules/common/megatron/rotary_pos_embedding.py
Comment (yidong72) [reply nemo/collections/nlp/modules/common/megatron/rotary_pos_embedding.py:29]: removed

File: nemo/collections/nlp/modules/common/megatron/transformer.py
Comment (yidong72) [reply nemo/collections/nlp/modules/common/megatron/transformer.py:1282]: Since the encoder_output is a tensor. The only difference between normal seq-to-seq encoder output and retrieval embedding is the order of dimension. One solution is to add this as an argument for `forward` method, but I guess it causes the problem of bloated arguments and need a few more if statements in the forward method.

File: nemo/collections/nlp/modules/common/megatron/transformer.py
Comment (yidong72) [reply nemo/collections/nlp/modules/common/megatron/transformer.py:392]: names are changed

File: nemo/collections/nlp/modules/common/megatron/transformer.py
Comment (yidong72) [reply nemo/collections/nlp/modules/common/megatron/transformer.py]: I added `retrieved_emb` as an argument.

**Review by ericharper — APPROVED** [review]
LGTM. Please wait for @MaximumEntropy 's approval as well.

**Review by MaximumEntropy — APPROVED** [review]

**Pull request merged**
